File: cxy/src/main/java/datastruct/btree/Page.java
package datastruct.btree;

import datastruct.array.Array;
import datastruct.array.Index;
import datastruct.array.SortedUniqArray;
import util.Tuple;

public class Page<K extends Comparable<K>>  {
  private int t;
  private int m;

  private SortedUniqArray<K> keys;
  private Array<Page<K>> children;

  public Page(int t) {
    this.t = t;
    this.m = 2 * t;
    keys = new SortedUniqArray<>(m - 1);
  }

  public K minKey() {
    return keys.first();
  }

  public K maxKey() {
    return keys.last();
  }

  public K get(int index) {
    return keys.get(index);
  }

  public Page<K> minChild() {
    return children == null ? null : children.first();
  }

  public Page<K> maxChild() {
    return children == null ? null : children.last();
  }

  public Page<K> getChild(int index) {
    return children == null ? null : children.get(index);
  }

  public boolean isFull() {
    return keys.isFull();
  }

  public boolean isLeaf() {
    return children == null;
  }

  public int size() {
    return keys.size();
  }

  /********************** 修改部分 *******************/
  /**
   * 只更新 key 的值
   */
  public K update(int index, K key) {
    K old = keys.get(index);
    keys.set(index, key);
    return old;
  }

  /** 针对叶子节点的增删操作 */
  public K addAsLeaf(K key) {
    keys.insert(key);
    return key;
  }

  public K removeAsLeaf(K key) {
    return keys.remove(key);
  }

  /**
   * 只有分裂 root 时才会被调用
   * @param child
   */
  public void addChild(Page<K> child) {
    if (this.children == null) {
      this.children = new Array<>(m);
    }
    children.insert(child);
  }

  /**
   * 新增一个key的情况，则需要新增一个child, 且插入在key之后
   * @param key
   * @param index
   * @param child
   */
  private void insertAt(K key, int index, Page<K> child) {
    keys.insertAt(index, key);
    children.insertAt(index+1, child);
  }

  public Tuple<K, Page<K>> remove(int index) {
    K key = keys.removeByIndex(index);
    Page<K> child = children == null ? null : children.removeByIndex(index+1);
    return Tuple.of(key, child);
  }

  /**
   * 查找指定的 key
   * @param key
   * @return if key exist, return the index of key, 否则返回 key 应该在的子节点的下标
   */
  public Index search(K key) {
    return keys.binarySearch(key);
  }

  /**
   * 1. 中间值 m 提升到父节点中
   * 2. 前面 t-1 个关键字和 t 个子节点作为 left
   * 3. 后面 t-1 个关键字和 t 个子节点作为 right
   * 4. left 仍然是父节点的子节点，且位置不动
   * 5. right 作为父节点的新子节点，插入在 indexAtParent + 1 的位置
   * @param parent
   * @param indexAtParent
   */
  public void split(Page<K> parent, int indexAtParent) {
    int t = this.t;
    int m = this.m;

    // 新生成一个节点
    Page<K> right = new Page<>(t);
    keys.moveTo(t, m-1, right.keys);
    if (this.children != null) {
      right.children = new Array<>(m);
      children.moveTo(t, m, right.children);
    }

    // 中间值 m 上提到父节点中
    K middle = keys.get(t-1);

    // 本节点作为left
    Page<K> left = this;
    left.keys.truncate(t-1);

    parent.insertAt(middle, indexAtParent, right);
  }

  public void merge(K middle, Page<K> right) {
    keys.insert(middle);
    right.keys.forEach(keys::insert);
    if (right.children != null) {
      right.children.forEach(this.children::insert);
    }
  }

  /**
   * 1. 如果子的左兄弟节点或者右兄弟节点至少有一个有 t 个元素，则从中借一个，以左兄弟为例：
   *  1. 将左兄弟节点的最大值上提值parent
   *  2. 将parent节点对应的关键字下降值 this
   *  3. 将左兄弟节点的最大child 作为 this 最小的child
   * 2. 如果都只有 t-1 个元素，则合并
   *
   * @param parent
   * @param index 自己作为子节点在parent.children中的下标
   * @return
   */
  public Page<K> borrowOrMergeFromSibling(Page<K> parent, int index) {
    // 1. 尝试从左兄弟借一个元素
    Page<K> left = null;
    if (index > 0 && (left = parent.children.get(index-1)).size() >= t) {
      K first = parent.get(index-1);
      K promot = left.keys.removeLast();
      parent.update(index-1, promot);
      this.keys.insertAt(0, first);
      // 同一层级，chilren如果不为空则都不为空
      if (this.children != null) {
        this.children.insertAt(0, left.children.removeLast());
      }
      return this;
    }

    // 2. 尝试从右兄弟借一个元素
    Page<K> right = null;
    if (index < parent.children.size() - 1 && (right = parent.children.get(index+1)).size() >= t) {
      K last = parent.get(index);
      K promot = right.keys.removeFirst();
      parent.update(index, promot);
      this.keys.insert(last);
      if (this.children != null) {
        this.children.insert(right.children.removeFirst());
      }
      return this;
    }

    // 3. 尝试与左兄弟合并, left, parent中下放一个, this
    if (left != null) {
      K middle = parent.keys.removeByIndex(index-1);
      // this 从 middle 的右节点删除
      parent.children.removeByIndex(index);
      left.merge(middle, this);
      return left;
    }

    if (right != null) {
      K middle = parent.keys.removeByIndex(index);
      parent.children.removeByIndex(index);
      this.merge(middle, right);
      // 替换一下
      parent.children.set(index, this);
      return this;
    }
    throw new RuntimeException("Something error.");
  }

  @Override
  public String toString() {
    return keys.toString();
  }

}

File: cxy/src/main/scala/tree/btree/Page.scala
package tree.btree

import collection.SortedUniqArray
import datastruct.array

class Page[K <: Comparable[K], V](val t: Int) {
  val m = t * 2
  val keys = new SortedUniqArray[K](m - 1)
  val values = new array.Array[V](m - 1)
  val children = new array.Array[Page[K, V]](m)

  /** read only operation */
  def isFull(): Boolean = keys.size() == m - 1

  def isLeaf(): Boolean = children.isEmpty

  def isNonLeaf(): Boolean = !isLeaf()

  def middleKey(): K = keys.get(keys.size() / 2)

  def find(key: K): (Boolean, Int) = keys.find(key)

  def size(): Int = keys.size()

  /** 更改 key, value */
  def put(key: K, value: V): Option[V] = {
    val (isExist, index) = keys.find(key)
    if (isExist) {
      update(key, value, index)
    } else {
      insertAt(key, value, index)
      None
    }
  }

  def insertAt(key: K, value: V, index: Int): Unit = {
    keys.insertAt(index, key)
    values.insertAt(index, value)
  }

  def update(key: K, value: V, index: Int): Option[V] = {
    val old = values.get(index)
    keys.set(index, key)
    values.set(index, value)
    Some(old)
  }

  def removeByIndex(index: Int): Option[V] = {
    val old = values.get(index)
    keys.removeByIndex(index)
    values.removeByIndex(index)
    Some(old)
  }



  def split(parent: Page[K, V], indexAtParent: Int): Page[K, V] = {
    require(isFull(), "Page is not full, but split: " + keys.size())
    val right = new Page[K, V](t)
    this.keys.moveTo(t, m-1, right.keys)
    this.values.moveTo(t, m-1, right.values)
    if (this.children.isNotEmpty) {
      this.children.moveTo(t + 1, m, right.children)
    }
    val middleKey = keys.removeLast()
    val middleValue = values.removeLast()

    if (parent == null) {
      val root = new Page[K, V](t)
      root.insertAt(middleKey, middleValue, indexAtParent)
      root.children.insert(this)
      root.children.insert(right)
      root
    } else {
      parent.insertAt(middleKey, middleValue, indexAtParent)
      parent.children.insertAt(indexAtParent+1, right)
      parent
    }
  }

  /**
   * 将 left, key, right 合并到 left 上
   * @param index
   * @return
   */
  def mergeIndexToLeft(index: Int): Page[K, V] = {
    val left = children.get(index)
    val right = children.removeByIndex(index+1)
    require(left.size() == t-1 && right.size() == t-1, s"Left[${left.size()}] or Right[${right.size()} is enough")
    val middleKey = keys.removeByIndex(index)
    val middleValue = values.removeByIndex(index)
    left.merge(middleKey, middleValue, right)
    left
  }

  def merge(middleKey: K, middleValue: V, right: Page[K, V]): Unit = {
    insertAt(middleKey, middleValue, size())
    val rightSize = right.size()
    right.keys.moveTo(0, rightSize, this.keys)
    right.values.moveTo(0, rightSize, this.values)
    if (right.children.isNotEmpty) {
      right.children.moveTo(0, rightSize+1, this.children)
    }
  }


  /**
   * 如果它的左兄弟有 t 个关键字，则从左兄弟借一个。
   * 如果它的右兄弟有 t 个关键字，则从右兄弟借一个。
   * 否则和它的一个兄弟节点合并
   * @param parent
   * @param indexAtParent
   */
  def borrowOrMergeFromSibling(parent: Page[K, V], indexAtParent: Int): Page[K, V] = {
    var left: Page[K, V] = null
    // 1. 尝试从左兄弟借一个
    if (indexAtParent > 0) {
      left = parent.children.get(indexAtParent-1)
      if (left.size() >= t) {
        // 父节点的关键下移
        keys.insertAt(0, parent.keys.get(indexAtParent-1))
        values.insertAt(0, parent.values.get(indexAtParent-1))
        // 左兄弟的最大关键字上提
        parent.keys.set(indexAtParent-1, left.keys.removeLast())
        parent.values.set(indexAtParent-1, left.values.removeLast())
        // 左兄弟的最大 child 作为自己的第一个 child
        if (children.isNotEmpty) {
          children.insertAt(0, left.children.removeLast())
        }
        return this
      }
    }

    var right: Page[K, V] = null
    // 2. 尝试从右兄弟借一个
    if (indexAtParent < parent.children.size() - 1) {
      right = parent.children.get(indexAtParent+1)
      if (right.size() >= t) {
        // 父节点的关键字下移
        keys.insertAt(keys.size(), parent.keys.get(indexAtParent))
        values.insertAt(values.size(), parent.values.get(indexAtParent))
        // 右兄弟的最小关键字上提
        parent.keys.set(indexAtParent, right.keys.removeFirst())
        parent.values.set(indexAtParent, right.values.removeFirst())
        // 右兄弟最小的 child 作为自己的第一个 child
        if (children.isNotEmpty) {
          children.insertAt(children.size(), right.children.removeFirst())
        }
        return this
      }
    }

    // 将自己和父节点的一个关键字合并到左子树上
    if (left != null) {
      return parent.mergeIndexToLeft(indexAtParent-1)
    }

    if (right != null) {
      return parent.mergeIndexToLeft(indexAtParent)
    }
    throw new RuntimeException("Something error.")
  }
}
